### ISSUE TITLE
SDL 2

### ISSUE THREAD

**Issue by Mako88** [dxx-rebirth/dxx-rebirth#82]
What are the pros and cons to upgrading to SDL 2? 1.2 is no longer supported and it may be time to move on...

**Comment by vLKp** [comment]
SDL2 changed joystick handling in a way that requires Linux users to grant permissions they may not want to grant.

**Comment by zicodxx** [comment]
It's true SDL2 changed a few ways of how Joysticks are handled, but it's not exclusively to blame at SDL2 itself. I worked on a few SDL2 projects so far and from all I can tell, all it needs is evdev support. The rest depends on the implementation. Lots of games want to go for the XBox 360 style layout and use mapping lists to support as many controllers as possible, relying on ID strings which will simply not work if you use some Third-Party controllers. This would of course not really matter for us as we have customizable controls.  

Maybe I am missing something but apart from that, I have not encountered too much of an issue with SDL2 and I can honestly not say how many distributions do not come with evdev supported by default. 

I would - however - like to start a n SDL2 branch after the release of the next version and start with the implementation so everyone can check out how it works for them.

Apart from that, I think SDL1.2 is still maintained and will be for a while. It will just not get any new features. But switching to SDL2 might help on multiple monitor support as well as the implementation of Force Feedback.

**Comment by vLKp** [comment]
[zicodxx wrote](https://github.com/dxx-rebirth/dxx-rebirth/issues/82#issuecomment-107923525):
> I would - however - like to start a n SDL2 branch after the release of the next version and start with the implementation so everyone can check out how it works for them.

I am fine with this.  Depending on how intrusive the porting process is, we might be able to maintain it as a compile-time selectable option and develop it on the main branch.

**Comment by zicodxx** [comment]
Yes, I think this should be possible via compile-time option. The current SDL code is fairly well abstracted, too and I'll try to keep this in order. The only addition I am planning is a few new functions for Force Feedback which will simply be dummies if compiled for SDL1.2x.

**Comment by vLKp** [comment]
I pushed [Add SConf option to use SDL2](/dxx-rebirth/dxx-rebirth/commit/a2cb5fea6ec07127134d98590d227cf08d59ed57) to add a new configure option `sdl2`.  When enabled, SConf will choose SDL2 instead of SDL1.  I pushed [SDL2: SDL_JoystickName parameter changed](/dxx-rebirth/dxx-rebirth/commit/19ed586988ffd4330f8cafd18b5f969535aab357) to fix a simple break caused by the new signature of `SDL_JoystickName`.  I pushed [SDL2: skip build of rbaudio.cpp](/dxx-rebirth/dxx-rebirth/commit/51ee9d2833fdf39da6aa52387887b3d71ea27c5f) to exclude building `rbaudio.cpp` in SDL2 mode, since it relies heavily on SDL CD-ROM support, and SDL2 removed that support.

Due to the many API changes in SDL2, the SDL2 build does not work yet.  However, changing the configure code to pick SDL2 should serve as a good starting point.

**Comment by zicodxx** [comment]
After the next release, I will have a lot of fun making that work. Especially looking forward to Force Feedback. Thanks vLKp. :)

**Comment by parkerlreed** [comment]
Just tried with `scons sdl2=1` and it bails when it gets to the mixer code.

```
...
...
CXX dxx . common/arch/ogl/ogl_sync.cpp
CXX dxx . common/arch/sdl/digi_mixer_music.cpp
common/arch/sdl/digi_mixer_music.cpp: In function ‘int dcx::mix_play_file(const char*, int, void (*)())’:
common/arch/sdl/digi_mixer_music.cpp:95:40: error: too few arguments to function ‘Mix_Music* Mix_LoadMUS_RW(SDL_RWops*, int)’
   current_music.reset(Mix_LoadMUS_RW(rw), rw);
                                        ^
In file included from common/arch/sdl/digi_mixer_music.cpp:17:0:
/usr/include/SDL2/SDL_mixer.h:168:37: note: declared here
 extern DECLSPEC Mix_Music * SDLCALL Mix_LoadMUS_RW(SDL_RWops *src, int freesrc);
                                     ^~~~~~~~~~~~~~
common/arch/sdl/digi_mixer_music.cpp:136:41: error: too few arguments to function ‘Mix_Music* Mix_LoadMUS_RW(SDL_RWops*, int)’
    current_music.reset(Mix_LoadMUS_RW(rw), rw);
                                         ^
In file included from common/arch/sdl/digi_mixer_music.cpp:17:0:
/usr/include/SDL2/SDL_mixer.h:168:37: note: declared here
 extern DECLSPEC Mix_Music * SDLCALL Mix_LoadMUS_RW(SDL_RWops *src, int freesrc);
                                     ^~~~~~~~~~~~~~
scons: *** [common/arch/sdl/digi_mixer_music.o] Error 1
scons: building terminated because of errors.
Failed target count: total=1; targets with enable_build_failure_summary=1: 1
Failed node list:
	common/arch/sdl/digi_mixer_music.o
Failed command list:
 g++ -c -g -O2 -ftabstop=4 -Wall -Wshadow -Werror=extra -Werror=format=2 -Werror=missing-braces -Werror=missing-include-dirs -Werror=uninitialized -Werror=undef -Werror=pointer-arith -Werror=cast-qual -Werror=missing-declarations -Werror=redundant-decls -Werror=vla -pthread -funsigned-char -std=gnu++14 -Werror=unused -Werror=useless-cast -fvisibility=hidden -Wsuggest-attribute=noreturn -Wlogical-op -Wold-style-cast -Wno-format-truncation -Wno-implicit-fallthrough -Wno-sign-compare -DPHYSFS_DEPRECATED= -DNDEBUG -DRELEASE -D_REENTRANT -Icommon/include -Icommon/main -I. -I/usr/include/SDL2 "common/arch/sdl/digi_mixer_music.cpp" -o "common/arch/sdl/digi_mixer_music.o"
```

**Comment by vLKp** [comment]
Yes, the Rebirth SDL2 build has never worked, as best I can recall.  The commits I described earlier in this thread were a few early steps toward it, but I never had much time or interest for it.  I only wrote as much as I did to save others the trouble of modifying the configure script.  There might be a bit more SDL2 related work than was cited here, but not much more, at least from me.  The configure knob exists solely to facilitate easy switching for people who want to develop SDL2 support in Rebirth.  If you are such a person, I'd be happy to review your changes.  Otherwise, this will likely wait until zico has time to work on it.

**Comment by parkerlreed** [comment]
@vLKp Ahh thanks. More of just a curious user. Glad to know.

**Comment by vLKp** [comment]
I see.  Might I interest you in becoming the person credited for porting to SDL2? 😉 

`sdl2` is marked `!EXPERIMENTAL!` in the help text, though I can see how that can mean different things to different people.  Some might interpret it to mean that the feature is functional but provided as-is (with no support from the developer).  Sadly, in this case, it's not even that far yet.  I could modify its help string or add other warning output if you have suggestions for how to more clearly communicate that this is not expected to work and no one should waste their time trying it.
